internal/gatewayapi/tls.go

```diff
@@ -11,51 +11,88 @@ import (
 	"fmt"
 
 	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
 )
 
 // validateTLSSecretData ensures the cert and key provided in a secret
 // is not malformed and can be properly parsed
-func validateTLSSecretData(secret *corev1.Secret) error {
-	certData := secret.Data[corev1.TLSCertKey]
+func validateTLSSecretsData(secrets []*corev1.Secret, host *v1beta1.Hostname) error {
+	var publicKeyAlgorithm string
+	var parseErr error
 
-	certBlock, _ := pem.Decode(certData)
-	if certBlock == nil {
-		return fmt.Errorf("unable to decode pem data in %s", corev1.TLSCertKey)
-	}
+	pkaSecretSet := make(map[string][]string)
+	for _, secret := range secrets {
+		certData := secret.Data[corev1.TLSCertKey]
 
-	_, err := x509.ParseCertificate(certBlock.Bytes)
-	if err != nil {
-		return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
-	}
+		certBlock, _ := pem.Decode(certData)
+		if certBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSCertKey)
+		}
 
-	keyData := secret.Data[corev1.TLSPrivateKeyKey]
+		cert, err := x509.ParseCertificate(certBlock.Bytes)
+		if err != nil {
+			return fmt.Errorf("unable to parse certificate in %s: %w", corev1.TLSCertKey, err)
+		}
+		publicKeyAlgorithm = cert.PublicKeyAlgorithm.String()
 
-	keyBlock, _ := pem.Decode(keyData)
-	if keyBlock == nil {
-		return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
-	}
+		keyData := secret.Data[corev1.TLSPrivateKeyKey]
 
-	var parseErr error
+		keyBlock, _ := pem.Decode(keyData)
+		if keyBlock == nil {
+			return fmt.Errorf("unable to decode pem data in %s", corev1.TLSPrivateKeyKey)
+		}
 
-	switch keyBlock.Type {
-	case "PRIVATE KEY":
-		_, err := x509.ParsePKCS8PrivateKey(keyBlock.Bytes)
+		matchedFQDN, err := verifyHostname(cert, host)
 		if err != nil {
-			parseErr = fmt.Errorf("unable to parse PKCS8 formatted private key in %s", corev1.TLSPrivateKeyKey)
+			return fmt.Errorf("hostname %s does not match Common Name or DNS Names in the certificate %s", string(*host), corev1.TLSCertKey)
```

Comment:
Yes, but with the change to the function signature for `validateTLSSecretsData` there is no guarantee that the error returned is associated with the secret for that listener. 

For example, lets say there are 3 `secrets` passed in and the first secret is bad then this function would always return the error for the first secret. When in theory the 3 secret is the one associated to the line you shared. 

My guess is the reason for the function signature change is so that you can add additional validation for checking for overlapping certs? Can this maybe be done independent of changing this func?

Reply:
Ah, I think you are right
Earlier after review, I did change `validateTLSSecretsData` func just to keep all the logic inside of this function.
Also in between there was added another TLS validation, so it looks a bit different now.
We could extract overlapping certs validation again, WDYT @arkodg ?

Reply:
looks like I missed this earlier - `validateTLSSecretsData` needs to be called outside the certificateRefs loop after collecting all `secrets` & you can embed the secret name/namespace within the err string within `validateTLSSecretsData`

Reply:
Sorry I'm on vacation so was slow to re-review. The final changes lgtm and it looks like it was already merged. Thanks 😊